src/walk-manifest.js

```diff
@@ -49,116 +50,197 @@ var parseManifest = function(content) {
-	}
-
-  manifest.content = syncRequest('GET', manifest.uri).getBody();
-  manifest.parsed  = parseManifest(manifest.content);
-	manifest.parsed.segments = manifest.parsed.segments   || [];
-	manifest.parsed.playlists = manifest.parsed.playlists || [];
-	manifest.parsed.mediaGroups = manifest.parsed.mediaGroups || {};
-
-  var playlists = manifest.parsed.playlists.concat(mediaGroupPlaylists(manifest.parsed.mediaGroups));
-	var key = parseKey(basedir, decrypt, resources, manifest, parent);
-
-	// SEGMENTS
-	manifest.parsed.segments.forEach(function(s, i) {
-		if (!s.uri) {
-			return;
-		}
-		// put segments in manifest-name/segment-name.ts
-		s.file = path.join(path.dirname(manifest.file), fsSanitize(path.basename(s.uri)));
-		if (!isAbsolute(s.uri)) {
-			s.uri = joinURI(path.dirname(manifest.uri), s.uri);
-		}
-		if (key) {
-			s.key = key;
-			s.key.iv = s.key.iv || new Uint32Array([0, 0, 0, manifest.parsed.mediaSequence, i]);
-		}
-
-		manifest.content = new Buffer(manifest.content.toString().replace(
-			s.uri,
-			path.relative(path.dirname(manifest.file), s.file)
-		));
-		resources.push(s);
-	});
-
-	// SUB Playlists
-	playlists.forEach(function(p, z) {
-		if (!p.uri) {
-			return;
-		}
-		resources = resources.concat(walkPlaylist(decrypt, basedir, p.uri, manifest, z));
-	});
-
-	return resources;
+var walkPlaylist = function(options, callback) {
+
+  var {
+    decrypt,
+    basedir,
+    uri,
+    parent = false,
+    manifestIndex = 0,
+    onError = function(err, resources, callback) {
+      // We get arrays back for sub playlist errors
+      if (err instanceof Array) {
+        // Just take first error we got a blow up
+        var e = err[0];
+        callback(e)
+      } else {
+        callback(err)
+      }
+    },
+    visitedUrls = []
+  } = options;
+
+  var resources = [];
+  var manifest = {};
+  manifest.uri = uri;
+  manifest.file = path.join(basedir, fsSanitize(path.basename(uri)));
+
+  // if we are not the master playlist
+  if (parent) {
+    manifest.file = path.join(
+      path.dirname(parent.file),
+      'manifest' + manifestIndex,
+      fsSanitize(path.basename(manifest.file))
+    );
+    // get the real uri of this playlist
+    if (!isAbsolute(manifest.uri)) {
+      manifest.uri = joinURI(path.dirname(parent.uri), manifest.uri);
+    }
+    // replace original uri in file with new file path
+    parent.content = new Buffer(parent.content.toString().replace(uri, path.relative(path.dirname(parent.file), manifest.file)));
+  }
+
+  if (visitedUrls.includes(manifest.uri)) {
+    const manifestError = new Error('Trying to visit the same uri again; stuck in a cycle|' + manifest.uri);
+    return onError(manifestError, resources, callback);
+  }
+
+  request({url: manifest.uri, timeout: 1500}, function(error, response, body) {
+    if (error) {
+      const manifestError = new Error(error.message + '|' + manifest.uri);
+      return onError(manifestError, resources, callback);
```

Comment:
I'd think for both of these passing the richer object through to the error handler in form or another may be better, especially so the handler can branch without having to do string parsing.